Kernel/NVMe: Minor fixes and improvements

Panky-codes opened the pull request:
This PR is a collection of some minor Fixes and an improvement in the NVMe driver. 
The only noticeable change is the addition of `NVMeSubmission` union.

IdanHo commented on Kernel/Storage/NVMe/NVMeController.cpp:
```diff
@@ -48,7 +48,7 @@ UNMAP_AFTER_INIT ErrorOr<void> NVMeController::initialize()
     m_controller_regs = TRY(Memory::map_typed_writable<volatile ControllerRegister>(PhysicalAddress(m_bar)));
```
`adds` -> `add` in the commit message

Panky-codes replied:
I assume you meant this ` all major OSs such as Linux, FreeBSD adds 1 to the` to ` all major OSs such as Linux, FreeBSD add 1 to the`

IdanHo replied:
right

Panky-codes replied:
Fixed it as well :)

IdanHo commented on Kernel/Storage/NVMe/NVMeDefinitions.h:
```diff
@@ -160,3 +160,71 @@ struct NVMeSubmission {
+    u64 rsvd12[2];
+};
+
+union NVMeSubmission {
+    NVMeGenericCmd generic;
+    NVMeIdentifyCmd identify;
+    NVMeRWCmd rw;
+    NVMeCreateCQCmd create_cq;
+    NVMeCreateSQCmd create_sq;
+};
```
This should be a struct with fields `op`, `flags`, `cmdid`, (assuming these are shared fields) and then this union instead, this deduplicates the definitions as well as allowing future code to decide which of the union members should be accessed based on these shared fields.

Panky-codes replied:
That's a good idea. Fixed it now.

IdanHo reviewed (commented):
Looks good, just a couple of nitpicks

IdanHo commented:
Note that this is failing to build on CI

Panky-codes commented:
@IdanHo Yeah, I noticed that. It is failing for x86_64 on a `static_assert` for the NVMeSubmission struct size. I check whether the size is 64 bytes. Somehow this is happening only for x86_64 target and not for i686. Any idea why as to why the size changes?

tomuta commented on Kernel/Storage/NVMe/NVMeController.cpp:
```diff
@@ -48,7 +48,7 @@ UNMAP_AFTER_INIT ErrorOr<void> NVMeController::initialize()
 
     auto caps = m_controller_regs->cap;
-    m_ready_timeout = Time::from_milliseconds(CAP_TO(caps) * 500); // CAP.TO is in 500ms units
+    m_ready_timeout = Time::from_milliseconds((CAP_TO(caps) + 1) * 500); // CAP.TO is in 500ms units
```
Wouldn't this then warrant removing the `min(1, ....)` from `wait_for_ready`?

Panky-codes replied:
That is correct! I will fix it up!

IdanHo commented:
> @IdanHo Yeah, I noticed that. It is failing to build for the x86_64 target on a `static_assert` for the NVMeSubmission struct size. I check whether the size is 64 bytes. Somehow this is happening only for x86_64 target and not for i686. Any idea as to why the size changes?

From a quick look, no, print the sizes of the sub-structs and see which field changes size/alignment.

Panky-codes commented:
@IdanHo I found the problem. As I removed 3 fields and moved it as a part of the common struct, the compiler was basically playing with padding on the `NVMe*Cmd` structs. Adding `[[gnu::packed]]` to the individual struct should fix the issue of the NVMeSubmission struct size being equal to 64 bytes. 
NOTE: the CI failing on i686 is due to some other issue : ` Unable to init server: Could not connect: Connection refused`

IdanHo commented on Kernel/Storage/NVMe/NVMeDefinitions.h:
```diff
@@ -160,3 +154,64 @@ struct NVMeSubmission {
+    u64 rsvd12[2];
+};
+
+struct NVMeSubmission {
```
Shouldn't this be marked `[[gnu::packed]]` as well?

Panky-codes replied:
I tried first that but the compiler ignores the `[[gnu::packed]]` attribute as we have a union. So I don't think it makes sense to have them. 
```
NVMeDefinitions.h:210:5: error: ignoring packed attribute because of unpacked non-POD field 'NVMeSubmission::<unnamed union> NVMeSubmission::<anonymous>'
```

Panky-codes replied:
Clang (https://godbolt.org/z/b3n459sae) doesn't emit a warning but gcc does (https://godbolt.org/z/Ys73x4dj3).

IdanHo replied:
You just need to declare the anonymous union as packed as well: https://godbolt.org/z/4r3hhE8d8

Panky-codes replied:
Somehow my primitive brain thought that it was not possible for the compiler to pack a union. Lol. I will fix it. Thanks!

IdanHo reviewed (approved):
Thanks
